snapcraft/plugins/_ros/rospack.py

```diff
@@ -0,0 +1,146 @@
+        self._ubuntu_keyrings = ubuntu_keyrings
+        self._project = project
+
+        self._rospack_install_path = os.path.join(self._rospack_path, "install")
+        self._rospack_cache_path = os.path.join(self._rospack_path, "cache")
+
+    def setup(self):
+        # Make sure we can run multiple times without error
+        os.makedirs(self._rospack_install_path, exist_ok=True)
+
+        # rospack isn't necessarily a dependency of the project, so we'll unpack
+        # it off to the side and use it from there.
+        logger.info("Preparing to fetch rospack...")
+        ubuntu = repo.Ubuntu(
+            self._rospack_path,
+            sources=self._ubuntu_sources,
+            keyrings=self._ubuntu_keyrings,
+            project_options=self._project,
+        )
+
+        logger.info("Fetching rospack...")
+        ubuntu.get(
+            [
+                "ros-{}-rospack".format(self._ros_distro),
+                "ros-{}-catkin".format(self._ros_distro),
+            ]
+        )
+
+        logger.info("Installing rospack...")
+        ubuntu.unpack(self._rospack_install_path)
+
+    def list_names(self):
+        """Obtain list of packages present in the workspace."""
+        output = self._run(["list-names"]).strip()
+        if output:
+            return set(output.split("\n"))
+        else:
+            return set()
+
+    def _run(self, arguments):
+        with tempfile.NamedTemporaryFile(mode="w+") as f:
+            lines = [
+                'export PYTHONPATH="{}"'.format(
+                    os.path.join(
+                        self._rospack_install_path,
+                        "usr",
+                        "lib",
+                        "python2.7",
+                        "dist-packages",
+                    )
+                )
+            ]
+
+            ros_path = os.path.join(
+                self._rospack_install_path, "opt", "ros", self._ros_distro
+            )
+            bin_paths = (
+                os.path.join(ros_path, "bin"),
+                os.path.join(self._rospack_install_path, "usr", "bin"),
+            )
+            lines.append(
+                "export {}".format(
+                    formatting_utils.format_path_variable(
+                        "PATH", bin_paths, prepend="", separator=":"
+                    )
+                )
+            )
+
+            lib_paths = common.get_library_paths(
+                self._rospack_install_path, self._project.arch_triplet
+            )
+            if lib_paths:
+                lines.append(
+                    "export {}".format(
+                        formatting_utils.format_path_variable(
+                            "LD_LIBRARY_PATH", lib_paths, prepend="", separator=":"
+                        )
+                    )
+                )
+
+            # Source our own workspace so we have all of rospack's dependencies
+            lines.append(
+                "_CATKIN_SETUP_DIR={} source {}".format(
```

Comment:
why are you sourcing this way if you are writing a script?
Aside from the need to source, why not construct the rest so that is passed on with `env`?

Reply:
Could perhaps use Popen with `env` and `shell=True` and drop this script wrapper?
e.g.
`Popen("source ...; do something", shell=True, env=env...)`

Reply:
`_CATKIN_SETUP_DIR` is only a temporary variable that we don't want to set beyond the sourcing of this script. This pattern is already well-established in other wrapper classes.